subiquity/server/controllers/snaplist.py

```diff
@@ -169,15 +199,24 @@ def make_autoinstall(self):
         return [attr.asdict(sel) for sel in self.model.selections]
 
     async def GET(self, wait: bool = False) -> SnapListResponse:
-        if self.loader.failed or not self.app.base_model.network.has_network:
+        if self.loader.fetch_list_failed() \
+                or not self.app.base_model.network.has_network:
             await self.configured()
             return SnapListResponse(status=SnapCheckState.FAILED)
-        if not self.loader.snap_list_fetched and not wait:
+        if not self.loader.fetch_list_completed() and not wait:
             return SnapListResponse(status=SnapCheckState.LOADING)
-        await self.loader.get_snap_list_task().wait()
-        if self.loader.failed or not self.app.base_model.network.has_network:
-            await self.configured()
-            return SnapListResponse(status=SnapCheckState.FAILED)
+        # Let's wait for the task to be completed.
+        # If the loader gets restarted, the cancellation should be absorbed.
+        while True:
+            await self.loader.load_list_task_created.wait()
+            try:
+                await self.loader.get_snap_list_task()
+                break
+            except SnapListFetchError:
+                await self.configured()
+                return SnapListResponse(status=SnapCheckState.FAILED)
+            except asyncio.CancelledError:
```

Comment:
Infinite retries on cancel + cancel being generally hard to get correct make me question this loop.
What's your thoughts on limiting it to some number of retries before giving up?

Reply:
What would you suggest as a number of retries?
IMHO, this is a good scenario where it makes sense to go for infinite retries on cancel ; since:
* the client explicitly provided the `wait=True` ; meaning they expect to wait
* the client provides a loading animation with a mechanism to skip the screen if the user does not want to wait longer
* cancellation on the server side should be the result of:
    * a network event
    * the app stopping - in which case the GET handler will be cancelled anyway

I think the worst case scenario would be a burst of network events that would keep the server busy-waiting. But the outcome would be the same with or without a limit in the GET handler.

Reply:
All right.  The log statement helps - if we see a long scroll of 'load list snap task ...' we will know roughly what happened.